server/iterable_weakmap.js

```diff
@@ -0,0 +1,227 @@
+/**
+ * @license
+ * IterableWeakMap
+ *
+ * Copyright 2018 Google Inc.
+ * https://github.com/NeilFraser/CodeCity
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview A WeakMap that's iterable.
+ * @author cpcallen@google.com (Christohper Allen)
+ */
+'use strict';
+
+const weak = require('weak');
+
+/**
+ * Function to clean up dead cells from an IterableWeakMap when a key
+ * object has been garbaged collected.
+ * @this {!IterableWeakMap} IterableWeakMap to remove GCed key from.
+ * @param {!Cell} cell Cell containing GCed key.
+ * @return {void}
+ */
+function cleanup(cell) {
+  this.cells_.delete(cell);
+}
+
+/**
+ * A weak-key, value tuple in an IterableWeakMap.
+ * @template KEY, VALUE
+ */
+class Cell {
+  /**
+   * @param {!IterableWeakMap} iwm Map this cell will belong to (for cleanup).
+   * @param {KEY} key The key for this cell.
+   * @param {VALUE} value The value for this cell.
+   */
+  constructor(iwm, key, value) {
+    /** @type {!WeakRef<KEY>} */
+    this.wk = weak(key, cleanup.bind(iwm, this));
+    /** @type {VALUE} */
+    this.value = value;
+  }
+
+  /**
+   * Return the cell's key (as a strong reference).
+   * @return {KEY}
+   */
+  getKey() {
+    return weak.get(this.wk);
+  }
+}
+
+/**
+ * A WeakMap implementing the full Map interface, including iterability.
+ * @struct
+ * @implements {Iterable<!Array<KEY|VALUE>>}
+ * @template KEY, VALUE
+ */
+// TODO(cpcallen): Make KEY a bounded to {!Object} once
+// closure-compiler supports bounded generic types
+class IterableWeakMap extends WeakMap {
+  /**
+    * @param {Iterable<!Array<KEY|VALUE>>|!Array<!Array<KEY|VALUE>>=} iterable
+   */
```

Comment:
Indent by one space.

Reply:
Actually, it's the previous line that needs outdenting.